### PR TITLE
Implement `rangeQuery` for VecSimTieredIndex - [MOD-5164]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

Implements `rangeQuery` for VecSimTieredIndex.

As part of this, the sorting (order) of the returned results moved from the C API to the C++ API and added an overload function for the range query call.

This PR also includes a fix for `searchBottomLayerEP`, where we (safely) get a copy of the old entry point, but by the time we get to get the `max_level_` it was updated to the new max level of the new entry point, and we tried to look at a non-existing neighbor list of the old entry point at the new max level.

**Main objects this PR modified**
1. range queries for `VecSimTieredIndex`
2. moved the sorting phase from C API into the C++ code
3. fix a bug in `searchBottomLayerEP`

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/unit/test_hnsw_tiered.cpp [hunk 1]
```diff
@@ -2799,3 +2799,305 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+TYPED_TEST(HNSWTieredIndexTest, RangeSearch) {
+    size_t dim = 4;
+    size_t k = 11;
+    size_t per_label = TypeParam::isMulti() ? 5 : 1;
+
+    size_t n_labels = k * 3;
+    size_t n = n_labels * per_label;
+
+    auto edge_delta = (k - 0.5) * per_label;
+    auto mid_delta = edge_delta / 2.1;
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:2812]: Why 2.1?

File: tests/unit/test_hnsw_tiered.cpp [hunk 2]
```diff
@@ -2799,3 +2799,305 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+
+    // Insert n/2 vectors to the flat index.
+    for (size_t i = (n + 1) / 2; i < n; i++) {
+        GenerateAndAddVector<TEST_DATA_T>(flat_index, dim, i / per_label, i);
+    }
+    ASSERT_EQ(tiered_index->indexSize(), n);
+    ASSERT_EQ(tiered_index->indexSize(), hnsw_index->indexSize() + flat_index->indexSize());
+
+    cur_memory_usage = allocator->getAllocationSize();
+    // Search for `range` so all the vectors will be from the flat index.
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:2924]:
```suggestion
    // Search for `range` so all the vectors will be from the HNSW index.
```

File: tests/unit/test_hnsw_tiered.cpp [hunk 3]
```diff
@@ -2799,3 +2799,305 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+    runRangeQueryTest(tiered_index, query_0, range, ver_res_0, k, BY_ID);
+    runRangeQueryTest(tiered_index, query_0, range, ver_res_0, k, BY_SCORE);
+    // Search for `range` so all the vectors will be from the main index.
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:2927]:
```suggestion
    // Search for `range` so all the vectors will be from the flat index.
```

File: tests/unit/test_hnsw_tiered.cpp [hunk 4]
```diff
@@ -2799,3 +2799,305 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+
+    auto *tiered_index = this->CreateTieredHNSWIndex(hnsw_params, &jobQ, index_ctx, &memory_ctx);
+    auto allocator = tiered_index->getAllocator();
+    EXPECT_EQ(index_ctx->index_strong_ref.use_count(), 1);
+
+    std::atomic_int successful_searches(0);
+    auto parallel_range_search = [](AsyncJob *job) {
+        auto *search_job = reinterpret_cast<SearchJobMock *>(job);
+        size_t k = search_job->k;
+        // size_t per_label = search_job->n;
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:3043]: remove

File: tests/unit/test_hnsw_tiered.cpp [hunk 5]
```diff
@@ -2799,3 +2799,305 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+        size_t dim = search_job->dim;
+        // The range that will get us k results.
+        double range = dim * (k - 0.5) * (k - 0.5) / 4; // L2 distance.
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:3046]: Explain the math

File: tests/unit/test_hnsw_tiered.cpp [hunk 6]
```diff
@@ -2799,3 +2799,305 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+
+    // Fill the job queue with insert and search jobs, while filling the flat index, before
+    // initializing the thread pool.
+    for (size_t i = 0; i < n; i++) {
+        // Insert a vector to the flat index and add a job to insert it to the main index.
+        GenerateAndAddVector<TEST_DATA_T>(tiered_index, dim, i % n_labels, i);
+
+        // Add a search job. Make sure the query element is between k and n - k.
+        auto query = (TEST_DATA_T *)allocator->allocate(dim * sizeof(TEST_DATA_T));
+        GenerateVector<TEST_DATA_T>(query, dim, ((n - i) % (n_labels - (2 * k))) + k);
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:3068]: Explain further the math...

**Review by alonre24 — COMMENTED** [review]
No actual comments, just add clarifications in tests

File: tests/unit/test_hnsw_tiered.cpp (hunk 1 above)
Comment (GuyAv46) [reply tests/unit/test_hnsw_tiered.cpp:2812]: an edge case in multi-value required some a bit higher value than 2

**Review by alonre24 — APPROVED** [review]

File: tests/unit/test_hnsw_tiered.cpp [hunk 7]
```diff
@@ -2799,3 +2799,316 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+    // Define the verification functions.
+    auto ver_res_0 = [&](size_t id, double score, size_t index) {
+        EXPECT_EQ(id, index);
+        // The expected score is the distance to the first vector of `id` label.
+        auto element = id * per_label;
+        EXPECT_DOUBLE_EQ(score, dim * element * element);
+    };
+
+    auto ver_res_1mid_by_id = [&](size_t id, double score, size_t index) {
+        size_t q_id = query_1mid[0] / per_label;
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_tiered.cpp:2859]: let's discuss those calculations, I think we can simplify them

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 8]
```diff
@@ -805,7 +800,6 @@ TieredHNSWIndex<DataType, DistType>::TieredHNSW_BatchIterator::getNextResults(
             auto tail = this->flat_iterator->getNextResults(
                 n_res - VecSimQueryResult_Len(this->flat_results), BY_SCORE_THEN_ID);
             concat_results(this->flat_results, tail);
-            VecSimQueryResult_Free(tail);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:808]: I figured it make more sense that `concat_results` will consume the second result array

File: tests/unit/test_hnsw_tiered.cpp [hunk 9]
```diff
@@ -718,7 +718,7 @@ TYPED_TEST(HNSWTieredIndexTest, parallelSearch) {
         .dim = dim,
         .metric = VecSimMetric_L2,
         .multi = isMulti,
-        .efRuntime = 20,
+        .efRuntime = n,
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnsw_tiered.cpp:721]: the low `ef` value caused the test to be a bit flaky. The higher value will ensure 100% recall every time

**Review by DvirDukhan — APPROVED** [review]
